xdsl/transforms/csl_stencil_handle_async_flow.py

```diff
@@ -0,0 +1,233 @@
+            return
+        if (
+            op.next_op
+            and isinstance(call_op := op.next_op, csl.CallOp)
+            and op.next_op.next_op
+            and isinstance(op.next_op.next_op, csl.ReturnOp)
+        ):
+            terminator = op.post_process.block.last_op
+            assert isinstance(terminator, csl_stencil.YieldOp)
+            rewriter.insert_op(call_op.clone(), InsertPoint.before(terminator))
+            rewriter.erase_op(call_op)
+            return
+        # todo there is more code here, split the code into a new func and call this
+
+
+@dataclass()
+class ConvertForLoopToCallGraphPass(RewritePattern):
+    """
+    Translates for-loop to csl functions, to handle async communications in the loop body.
+    Splits the body of the enclosing function into: pre_block, loop, post_block.
+    Further splits `loop` into for_cond, for_body, for_inc.
+    Loop var and iter_args are promoted to module-level csl vars.
+
+    Limitations:
+      * Loop can only yield its input iter_args (in any order)
+      * Loop bounds and step must be arith.constant values
+    """
+
+    counter: int
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: scf.For, rewriter: PatternRewriter, /):
+        if not self._is_inside_wrapper_outside_apply(op):
+            return
+
+        parent_func = op.parent_op()
+        while parent_func is not None and not isinstance(parent_func, csl.FuncOp):
+            parent_func = parent_func.parent_op()
+        if not parent_func:
+            return
+
+        # limitation: can yield iter_args in any order, but they cannot be modified in the loop body
+        terminator = op.body.block.last_op
+        assert isinstance(terminator, scf.Yield)
+        assert all(
+            arg in op.body.block.args for arg in terminator.arguments
+        ), "Can only yield unmodified iter_args (in any order)"
+
+        # limitation: currently only loops built from arith.constant are supported
+        assert isinstance(op.lb, OpResult)
+        assert isinstance(op.ub, OpResult)
+        assert isinstance(op.step, OpResult)
+        assert isinstance(op.lb.op, arith.Constant)
+        assert isinstance(op.ub.op, arith.Constant)
+        assert isinstance(op.step.op, arith.Constant)
+        assert isa(op.lb.op.value, IntegerAttr[IndexType])
+        assert isa(op.ub.op.value, IntegerAttr[IndexType])
+        assert isa(op.step.op.value, IntegerAttr[IndexType])
+
+        # limitation: all iter_args must be memrefs (stencil buffers) and have the same data type
+        assert isa(op.iter_args[0].type, MemRefType[csl.ZerosOp.T])
+        element_type = op.iter_args[0].type.get_element_type()
+        assert all(
+            isa(a.type, MemRefType[csl.ZerosOp.T])
+            and element_type == a.type.get_element_type()
+            for a in op.iter_args
+        )
+
+        no_params = FunctionType.from_lists([], [])
+        cond_task_id = self.counter + 1
```

Comment:
Are you sure that this is a robust enough solution? I remember us thinking about having task-id allocation and referring to them by symbols in the meantime. What happened to that plan? (jsut curious, not necessarily criticism)

Reply:
Yes and no, with our current comms library we have ids 1-3 free for local task ids. We could prob put an assert not to exceed that, but I think it'll simply not compile at a later stage. The question is how much knowledge this pass should have of things downstream? That said, it is the first thing in the pipeline to handle tasks and task ids, so it has a bit of freedom.